raster-foundry/blasterjs
# Per-component themeing

## Overview

Refactor theme system to support per-component themeing.

Replaced `defaultTheme.js` with a `theme` folder:

```
packages/core/theme/
├── components/
│   ├── foo/
│       └── index.js
│   ├── bar/
│       └── index.js
│   └── index.js
├── base.js
├── icons.js
└── index.js
```

What was `defaultTheme.js` is now `base.js` and, as before, contains component-agnostic theme settings organized into styled-system theme objects.

`icons.js` remains untouched. I updated the CLI to its new location.

`index.js` compiles and exports `base.js`, `icons.js`, and the component theme files as a single `theme` object, with per-component theme properties distributed across their styled-system theme objects. Any references in the component themes to keys in `base.js` get resolved along the way.

For example, this `base.js`...

```js
// base.js
{
  space: [
    "0",
    "0.8rem",
    "1.6rem",
    ...
  ],
  colors: {
    white: "#ffffff",
    black: "#000000",
    primary: "#738FFC",
    secondary: "#2FC1D0",
    ...
  },
  ...
}
```

... and these component themes ...

```js
// components/foo.js

{
  space: {
    px: 2,
    py: 1
  },
  colors: {
    color: "primary",
    bg: "white"
  },
  ...
}
```

```js
// components/bar.js

{
  space: {
    m: 1,
    p: 1
  },
  colors: {
    color: "black",
    highlight: "secondary"
  },
  ...
}
```
result in this `theme` object:

```js
{
  space: {
    0: 0,
    1: "0.8rem",
    2: "1.6rem",
    ...
    foo: {
      px: "1.6rem",
      py: "0.8rem"
    },
    bar: {
      m: "0.8rem",
      p: "0.8rem"
    }
  },
  colors: {
    white: "#ffffff",
    black: "#000000",
    primary: "#738FFC",
    secondary: "#2FC1D0",
    ...
    foo: {
      color: "#738FFC",
      bg: "#ffffff"
    },
    bar: {
      color: "#000000",
      highlight: "#2FC1D0"
    }
  },
  ...
}
```


### Checklist


Are there any of the following in this PR?


If one of the above is checked...

- [x] information or guidance around needed changes for consumers of this library has been added to the changelog under `Upgrade Instructions`


### Notes

I had to bump `styled-system` to `3.2.1`.

The CLI needs to be updated to automatically generate (or destroy) the corresponding subfolder of `theme/components/`, and update `theme/components/index.js`, when `blaster g` (or `blaster d`) is run.


## Testing Instructions

 * Fiddle with values in `base.js`, make sure changes propagate to components that refer to those keys in their component themes.
 * Fiddle with values in a component's theme files. Use keys from `base.js` as well as valid strings or numbers that the corresponding props expect.
